src/system/SystemLayer.h

```diff
@@ -298,13 +326,22 @@ class LayerSocketsLoop : public LayerSockets
     virtual void HandleEvents()    = 0;
     virtual void EventLoopEnds()   = 0;
 
+#if !CHIP_SYSTEM_CONFIG_USE_DISPATCH
+    virtual void AddLoopHandler(EventLoopHandler & handler)    = 0;
+    virtual void RemoveLoopHandler(EventLoopHandler & handler) = 0;
+#endif // !CHIP_SYSTEM_CONFIG_USE_DISPATCH
+
 #if CHIP_SYSTEM_CONFIG_USE_DISPATCH
     virtual void SetDispatchQueue(dispatch_queue_t dispatchQueue) = 0;
     virtual dispatch_queue_t GetDispatchQueue()                   = 0;
 #elif CHIP_SYSTEM_CONFIG_USE_LIBEV
     virtual void SetLibEvLoop(struct ev_loop * aLibEvLoopP) = 0;
     virtual struct ev_loop * GetLibEvLoop()                 = 0;
 #endif // CHIP_SYSTEM_CONFIG_USE_DISPATCH/LIBEV
+
+protected:
+    // Expose EventLoopHandler.mState to sub-classes
+    decltype(EventLoopHandler::mState) & LoopHandlerState(EventLoopHandler & handler) { return handler.mState; }
```

Comment:
Why doesn't the EventLoopHandler provide a `GetState() const`? Seems safe enough.

Reply:
It's not const because the event loop implementations needs to write as well as read it. By using the friend class and this protected helper I'm making it only available to `LayerSocketsLoop` sub-classes, without having to list implementation classes explicitly as friends.

Reply:
This is similar to how `IntrusiveListNodePrivateBase` contains state (`mPrev` and `mNext`) for use by `IntrusiveList`, except here `LayerSocketsLoop` is abstract, so the state is really used by the concrete implementation class.